### PR TITLE
*: activate txn for query on infoschema tables

### LINKED ISSUE #57874: GC is wrongly triggered in dumpling+ million tables scenario
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

millions tables
dumpling dump a database with 2 tables
```
tiup dumpling:nightly -u root -P 4000 -h [10.104.100.98](http://10.104.100.98/) --filetype csv -t 16 -r 200000 -F 256MiB --database sb10wnormal1
The component `dumpling` version v8.5.0-alpha-nightly is not installed; downloading from repository.
download https://tiup-mirrors.pingcap.com/dumpling-v8.5.0-alpha-nightly-linux-amd64.tar.gz 37.20 MiB / 37.20 MiB 100.00% 44.63 MiB/s
Starting component dumpling: /root/.tiup/components/dumpling/v8.5.0-alpha-nightly/dumpling -u root -P 4000 -h [10.104.100.98](http://10.104.100.98/) --filetype csv -t 16 -r 200000 -F 256MiB --database sb10wnormal1
Release version: v8.5.0-alpha-258-g6d7407156c
Git commit hash: 6d7407156c17ce9785c91d716a640d641feafc7a
Git branch:      HEAD
Build timestamp: 2024-12-02 05:17:01Z
Go version:      go version go1.23.3 linux/amd64

[2024/12/02 06:50:54.598 +00:00] [INFO] [versions.go:54] ["Welcome to dumpling"] ["Release Version"=v8.5.0-alpha-258-g6d7407156c] ["Git Commit Hash"=6d7407156c17ce9785c91d716a640d641feafc7a] ["Git Branch"=HEAD] ["Build timestamp"="2024-12-02 05:17:01"] ["Go Version"="go version go1.23.3 linux/amd64"]
[2024/12/02 06:50:54.600 +00:00] [INFO] [version.go:457] ["detect server version"] [type=TiDB] [version=8.5.0-alpha-256-gec288d95a3]
[2024/12/02 06:50:54.631 +00:00] [INFO] [pd_service_discovery.go:1025] ["[pd] switch leader"] [new-leader=http://tc-pd-0.tc-pd-peer.forge2-tps-7690491-1-273.svc:2379/] [old-leader=]
[2024/12/02 06:50:54.631 +00:00] [INFO] [pd_service_discovery.go:499] ["[pd] init cluster id"] [cluster-id=7441598078004266891]
[2024/12/02 06:50:54.631 +00:00] [INFO] [client.go:532] ["[pd] changing service mode"] [old-mode=UNKNOWN_SVC_MODE] [new-mode=PD_SVC_MODE]
[2024/12/02 06:50:54.631 +00:00] [INFO] [tso_client.go:296] ["[tso] switch dc tso global allocator serving url"] [dc-location=global] [new-url=http://tc-pd-0.tc-pd-peer.forge2-tps-7690491-1-273.svc:2379/]
[2024/12/02 06:50:54.632 +00:00] [INFO] [client.go:538] ["[pd] service mode changed"] [old-mode=UNKNOWN_SVC_MODE] [new-mode=PD_SVC_MODE]
[2024/12/02 06:50:54.632 +00:00] [INFO] [tso_dispatcher.go:140] ["[tso] start tso deadline watcher"] [dc-location=global]
[2024/12/02 06:50:54.632 +00:00] [INFO] [tso_dispatcher.go:198] ["[tso] tso dispatcher created"] [dc-location=global]
[2024/12/02 06:50:54.632 +00:00] [INFO] [tso_client.go:132] ["[tso] start tso dispatcher check loop"]
[2024/12/02 06:50:54.632 +00:00] [INFO] [tso_dispatcher.go:476] ["[tso] start tso connection contexts updater"] [dc-location=global]
[2024/12/02 06:50:54.632 +00:00] [INFO] [tso_dispatcher.go:723] ["[tso] switching tso rpc concurrency"] [old=0] [new=1]
[2024/12/02 06:50:54.632 +00:00] [INFO] [dump.go:1523] ["generate dumpling gc safePoint id"] [id=dumpling_1733122254632766329]
[2024/12/02 06:50:54.641 +00:00] [INFO] [dump.go:152] ["begin to run Dump"] [conf="{\"s3\":{\"endpoint\":\"\",\"region\":\"\",\"storage-class\":\"\",\"sse\":\"\",\"sse-kms-key-id\":\"\",\"acl\":\"\",\"access-key\":\"\",\"secret-access-key\":\"\",\"session-token\":\"\",\"provider\":\"\",\"force-path-style\":true,\"use-accelerate-endpoint\":false,\"role-arn\":\"\",\"external-id\":\"\",\"object-lock-enabled\":false},\"gcs\":{\"endpoint\":\"\",\"storage-class\":\"\",\"predefined-acl\":\"\",\"credentials-file\":\"\"},\"azblob\":{\"endpoint\":\"\",\"account-name\":\"\",\"account-key\":\"\",\"access-tier\":\"\",\"sas-token\":\"\",\"encryption-scope\":\"\",\"encryption-key\":\"\"},\"SpecifiedTables\":false,\"AllowCleartextPasswords\":false,\"SortByPk\":true,\"NoViews\":true,\"NoSequences\":true,\"NoHeader\":false,\"NoSchemas\":false,\"NoData\":false,\"CompleteInsert\":false,\"TransactionalConsistency\":true,\"EscapeBackslash\":true,\"DumpEmptyDatabase\":true,\"PosAfterConnect\":false,\"CompressType\":0,\"Host\":\"[10.104.100.98](http://10.104.100.98/)\",\"Port\":4000,\"Threads\":16,\"User\":\"root\",\"Security\":{\"CAPath\":\"\",\"CertPath\":\"\",\"KeyPath\":\"\"},\"LogLevel\":\"info\",\"LogFile\":\"\",\"LogFormat\":\"text\",\"OutputDirPath\":\"./export-2024-12-02T06:50:54Z\",\"StatusAddr\":\":8281\",\"Snapshot\":\"454327600314580994\",\"Consistency\":\"snapshot\",\"CsvNullValue\":\"\\\\N\",\"SQL\":\"\",\"CsvSeparator\":\",\",\"CsvDelimiter\":\"\\\"\",\"CsvLineTerminator\":\"\\r\\n\",\"Databases\":[\"sb10wnormal1\"],\"Where\":\"\",\"FileType\":\"csv\",\"ServerInfo\":{\"ServerType\":3,\"ServerVersion\":\"8.5.0-alpha-256-gec288d95a3\",\"HasTiKV\":true},\"Rows\":200000,\"ReadTimeout\":900000000000,\"TiDBMemQuotaQuery\":0,\"FileSize\":268435456,\"StatementSize\":1000000,\"SessionParams\":{\"tidb_snapshot\":\"454327600314580994\"},\"Tables\":{},\"CollationCompatible\":\"loose\",\"CsvOutputDialect\":0,\"IOTotalBytes\":null,\"Net\":\"\"}"]
[2024/12/02 07:05:42.660 +00:00] [INFO] [resource_manager_client.go:296] ["[resource manager] exit resource token dispatcher"]
[2024/12/02 07:05:42.660 +00:00] [INFO] [tso_stream.go:359] ["tsoStream.recvLoop ended"] [stream=[tc-pd-0.tc](http://tc-pd-0.tc/)-pd-peer.forge2-tps-7690491-1-273.svc:2379-1] [error="rpc error: code = Canceled desc = context canceled"] [errorVerbose="rpc error: code = Canceled desc = context canceled\[ngithub.com/tikv/pd/client.(*tsoStream).recvLoop\n\tgithub.com/tikv/pd/client@v0.0.0-20241111073742-238d4d79ea31/tso_stream.go:427\nruntime.goexit\n\truntime/asm_amd64.s:1700](http://ngithub.com/tikv/pd/client.(*tsoStream).recvLoop/n/tgithub.com/tikv/pd/client@v0.0.0-20241111073742-238d4d79ea31/tso_stream.go:427/nruntime.goexit/n/truntime/asm_amd64.s:1700)"]
[2024/12/02 07:05:42.661 +00:00] [INFO] [tso_dispatcher.go:490] ["[tso] exit tso connection contexts updater"] [dc-location=global]
[2024/12/02 07:05:42.660 +00:00] [INFO] [tso_dispatcher.go:264] ["[tso] stop fetching the pending tso requests due to context canceled"] [dc-location=global]
[2024/12/02 07:05:42.661 +00:00] [INFO] [tso_dispatcher.go:201] ["[tso] exit tso dispatcher"] [dc-location=global]
[2024/12/02 07:05:42.661 +00:00] [INFO] [pd_service_discovery.go:551] ["[pd] exit member loop due to context canceled"]
[2024/12/02 07:05:42.661 +00:00] [INFO] [tso_client.go:147] ["[tso] exit tso dispatcher check loop"]
[2024/12/02 07:05:42.661 +00:00] [INFO] [tso_dispatcher.go:158] ["[tso] exit tso deadline watcher"] [dc-location=global]
[2024/12/02 07:05:42.661 +00:00] [ERROR] [main.go:78] ["dump failed error stack info"] [error="sql: SELECT COUNT(1) as c FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_TYPE='SEQUENCE': Error 9006 (HY000): GC life time is shorter than transaction duration, transaction starts at 2024-12-02 06:50:38.718 +0000 UTC, GC safe point is 2024-12-02 06:50:54.618 +0000 UTC"] [errorVerbose="Error 9006 (HY000): GC life time is shorter than transaction duration, transaction starts at 2024-12-02 06:50:38.718 +0000 UTC, GC safe point is 2024-12-02 06:50:54.618 +0000 UTC\nsql: SELECT COUNT(1) as c FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_TYPE='SEQUENCE'\[ngithub.com/pingcap/tidb/dumpling/export.CheckIfSeqExists\n\tgithub.com/pingcap/tidb/dumpling/export/sql.go:932\ngithub.com/pingcap/tidb/dumpling/export.prepareTableListToDump\n\tgithub.com/pingcap/tidb/dumpling/export/dump.go:1153\ngithub.com/pingcap/tidb/dumpling/export.(*Dumper).Dump\n\tgithub.com/pingcap/tidb/dumpling/export/dump.go:218\nmain.main\n\t./main.go:75\nruntime.main\n\truntime/proc.go:272\nruntime.goexit\n\truntime/asm_amd64.s:1700](http://ngithub.com/pingcap/tidb/dumpling/export.CheckIfSeqExists/n/tgithub.com/pingcap/tidb/dumpling/export/sql.go:932/ngithub.com/pingcap/tidb/dumpling/export.prepareTableListToDump/n/tgithub.com/pingcap/tidb/dumpling/export/dump.go:1153/ngithub.com/pingcap/tidb/dumpling/export.(*Dumper).Dump/n/tgithub.com/pingcap/tidb/dumpling/export/dump.go:218/nmain.main/n/t./main.go:75/nruntime.main/n/truntime/proc.go:272/nruntime.goexit/n/truntime/asm_amd64.s:1700)"]

dump failed: sql: SELECT COUNT(1) as c FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_TYPE='SEQUENCE': Error 9006 (HY000): GC life time is shorter than transaction duration, transaction starts at 2024-12-02 06:50:38.718 +0000 UTC, GC safe point is 2024-12-02 06:50:54.618 +0000 UTC
```
### 4. What is your TiDB version? (Required)

master

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close #57874

Problem Summary:

The GC should be blocked by query.

### What changed and how does it work?

There are two places that the transaction should blcok GC:

- The query "SELECT COUNT(1) as c FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_TYPE='SEQUENCE'" itself
- The internal transaction `meta.ListTables` invoked by `SchemaTableInfos()` API

This commit just fix the first one. It should be enough to fix issue 57874.
The root cause is that whe we query on infoschema tables, the transaction is not activated.
So the session process info would not be collected to block GC.

The second issue is caused by the fact that the meta API call does not block GC.
I don't take investigation on it further more, meta API is limited used.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [X] Manual test (add detailed scripts or steps below)

Modify the code and build tidb, check no more "GC life time is shorter than transaction duration" error.

```
diff --git a/pkg/planner/core/memtable_infoschema_extractor.go b/pkg/planner/core/memtable_infoschema_extractor.go
index c527dde968..34cabb3677 100644
--- a/pkg/planner/core/memtable_infoschema_extractor.go
+++ b/pkg/planner/core/memtable_infoschema_extractor.go
@@ -15,6 +15,7 @@
 package core
 
 import (
+       "time"
        "bytes"
        "context"
        "fmt"
@@ -184,6 +185,7 @@ func (e *InfoSchemaBaseExtractor) ListSchemasAndTables(
                return findTableAndSchemaByName(ctx, is, schemas, tableNames)
        }
        schemas := e.ListSchemas(is)
+       time.Sleep(12 * time.Minute)
        return listTablesForEachSchema(ctx, e, is, schemas)
 }
```

```
mysql> set @@tidb_mem_quota_query = 12*1024*1024*1024;
Query OK, 0 rows affected (0.00 sec)

mysql> select count(tidb_table_id) from information_schema.tables where table_name != 'test';
+----------------------+
| count(tidb_table_id) |
+----------------------+
|              1000818 |
+----------------------+
1 row in set (12 min 28.84 sec)
```

- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by lance6716 — APPROVED** [review]

**Comment by tiancaiamao** [comment]
The second case is still risky after this fix.

If we take a latest infoschema, and use it, continue for 10min (or maybe less)
The GC error may still happen because the internal txn does not block GC.

File: pkg/executor/memtable_reader.go [hunk 1]
```diff
@@ -93,6 +93,22 @@ func (*MemTableReaderExec) isInspectionCacheableTable(tblName string) bool {
...
+func (e *MemTableReaderExec) Open(ctx context.Context) error {
+	err := e.BaseExecutor.Open(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Activate the transaction, otherwise SELECT .. FROM INFORMATION_SCHEMA.XX .. does not block GC worker.
+	// And if the query last too long (10min), it causes error "GC life time is shorter than transaction duration"
+	if !strings.HasPrefix(e.table.Name.L, "cluster_") {
+		// Skip cluster_ tables because it could be a tidb coprocessor request and txn is not prepared.
```
Comment (lance6716) [inline pkg/executor/memtable_reader.go:106]: The special checking is an indication that `MemTableReaderExec` should be separated into multiple structure. As the name `MemTableReaderExec` I think it's a pure memory implementation, and no need to consider KV transaction, GC and other things. In infoschema v2 project, `MemTableReaderExec` involves KV read/write. And as this comment said, it also involves coprocessor request. In future we'd better separate InfoSchemaV2Executor and ClusterTablesExecutor.

**Review by lance6716 — COMMENTED** [review]
Can you find the place that min TS of active txn ? We can write a testkit test to check the minTS really takes effect with this PR

File: pkg/executor/memtable_reader.go (hunk 1 above)
Comment (tiancaiamao) [reply pkg/executor/memtable_reader.go]: This still fails to pass CI, there are more cases.
It's *toooooo difficult* to figure out all the callers of this function that `e.Ctx().Txn(true)` fails.

Comment (tiancaiamao) [reply pkg/executor/memtable_reader.go]: I change to another way to check, hope this time no panics.

**Review by wjhuang2016 — APPROVED** [review]

**Comment by tiancaiamao** [comment]
Create an issue https://github.com/pingcap/tidb/issues/57952 to trace the second problem.
That one is not blocking, I list as en enhancement, as most of the cases should have been avoided by this commit.

**Pull request merged**